server/mongodb/scheduler.go

```diff
@@ -0,0 +1,88 @@
+package mongodb
+
+import (
+	"fmt"
+	"github.com/ohsu-comp-bio/funnel/compute/scheduler"
+	pbs "github.com/ohsu-comp-bio/funnel/proto/scheduler"
+	"github.com/ohsu-comp-bio/funnel/proto/tes"
+	"golang.org/x/net/context"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/codes"
+	mgo "gopkg.in/mgo.v2"
+	"gopkg.in/mgo.v2/bson"
+)
+
+// QueueTask adds a task to the scheduler queue.
+func (db *MongoDB) QueueTask(task *tes.Task) error {
+	return nil
+}
+
+// ReadQueue returns a slice of queued Tasks. Up to "n" tasks are returned.
+func (db *MongoDB) ReadQueue(n int) []*tes.Task {
+	var tasks []*tes.Task
+	err := db.tasks.Find(bson.M{"state": tes.State_QUEUED}).Select(basicView).All(&tasks)
+	if err != nil {
+		fmt.Println(err)
+		return nil
+	}
+	return tasks
+}
+
+// PutNode is an RPC endpoint that is used by nodes to send heartbeats
+// and status updates, such as completed tasks. The server responds with updated
+// information for the node, such as canceled tasks.
+func (db *MongoDB) PutNode(ctx context.Context, node *pbs.Node) (*pbs.PutNodeResponse, error) {
+	var existing pbs.Node
+	err := db.nodes.Find(bson.M{"id": node.Id}).One(&existing)
+	if err != nil && err != mgo.ErrNotFound {
+		return nil, err
+	}
+
+	if existing.GetVersion() != 0 && node.Version != existing.GetVersion() {
+		return nil, fmt.Errorf("Version outdated")
+	}
+
+	err = scheduler.UpdateNode(ctx, db, node, &existing)
+	if err != nil {
+		return nil, err
+	}
+
+	_, err = db.nodes.Upsert(bson.M{"id": node.Id}, node)
+
+	return &pbs.PutNodeResponse{}, err
+}
+
+// GetNode gets a node
+func (db *MongoDB) GetNode(ctx context.Context, req *pbs.GetNodeRequest) (*pbs.Node, error) {
+	var node pbs.Node
+	err := db.nodes.Find(bson.M{"id": req.Id}).One(&node)
+	if err != nil {
+		if err == mgo.ErrNotFound {
```

Comment:
Could this error check be the same `DeleteNode` below?

Reply:
sure